File: frontend/src/lib/breadcrumbs.ts
// Stage 7 client-side breadcrumb ring buffer (D-A7). Last 20 events (route changes,
// scan submits, API errors, notification-panel opens), attached to a support ticket so
// admin can debug blind. RED LINE: only safe app metadata is ever recorded here, never a
// journal outcome value (the client does not hold unrevealed values; the server also
// re-sanitizes on write). Pure logic, unit-tested.

export interface Breadcrumb {
  event_type: string; // route_change | scan_submit | api_error | notif_open | ...
  ts: string; // ISO 8601
  path?: string;
  code?: number; // HTTP status for api_error
}

export const MAX_BREADCRUMBS = 20;

/** Append to a fixed-size ring buffer, keeping the newest `max` entries. Pure. */
export function ringPush<T>(list: T[], item: T, max: number): T[] {
  return [...list, item].slice(-max);
}

let buffer: Breadcrumb[] = [];

export function addBreadcrumb(event_type: string, extra?: { path?: string; code?: number }): void {
  buffer = ringPush(buffer, { event_type, ts: new Date().toISOString(), ...extra }, MAX_BREADCRUMBS);
}

/** Oldest-first snapshot (chronological), for the ticket payload / admin timeline. */
export function getBreadcrumbs(): Breadcrumb[] {
  return [...buffer];
}

export function clearBreadcrumbs(): void {
  buffer = [];
}

File: backend/core/breadcrumbs.py
"""Server-side breadcrumb sanitizer (Stage 7, D-A7 / red line).

Breadcrumbs are the reporter's last client-side events (route changes, scan submits, API
errors, notification-panel opens) attached to a support ticket. They are ALLOWLISTED
metadata only: this sanitizer drops every field not on the allowlist, so a journal
outcome value (status / r_result / win-loss) can never ride into a ticket payload — the
reveal-gating red line applies here too. The client never holds unrevealed values, and
this is the belt-and-suspenders server guard.
"""
from typing import Any

# Only these keys survive per crumb. Deliberately excludes anything outcome-shaped
# (status, r_result, outcome, pnl, direction, entry, sl, tp, score, ...).
_ALLOWED_KEYS = {
    "type", "event_type", "path", "route", "label", "ts", "timestamp", "code",
    "status_code", "method",
}
_MAX_ITEMS = 20
_MAX_STR = 200


def sanitize_breadcrumbs(raw: Any, max_items: int = _MAX_ITEMS) -> list[dict]:
    """Return an allowlisted, length-capped copy. Non-list / malformed input → []."""
    if not isinstance(raw, list):
        return []
    out: list[dict] = []
    for item in raw[:max_items]:
        if not isinstance(item, dict):
            continue
        clean: dict = {}
        for key, val in item.items():
            if key not in _ALLOWED_KEYS:
                continue
            if isinstance(val, bool):
                clean[key] = val
            elif isinstance(val, (int, float)):
                clean[key] = val
            elif isinstance(val, str):
                clean[key] = val[:_MAX_STR]
            # anything else (dict/list/None) is dropped
        if clean:
            out.append(clean)
    return out
